Readme.md
# Golive [![Build Status](https://travis-ci.org/alexjomin/golive.svg?branch=master)](https://travis-ci.org/alexjomin/golive)

## Description

With **golive**, keep track of your deliveries.

### what, when and where ?

In a continious deployment workflow, it's sometimes difficult to remember **which** software has been deployed, **when** and **where**. golive offers a dead simple api to record and retrieve your deliveries, it also come with a frontend to display the timeline with a heatmap.

## Screenshot

![screenshot](etc/screenshot.png)

## Features

* Dead simple HTTP API, written in Go.
* Web dashboard based on Vue.js and Bootstrap :
	* Timeline of your deployments
	* List of the softwares, with filter.
	* List of the environments, with filter.
	* HeatMap, github style.
* Included database : [Bolt](https://github.com/boltdb/bolt).
* Super easy to install !

## Install
This project uses [Glide](https://github.com/Masterminds/glide), a Vendor Package Management
	
	glide install
	make build

## Start

	golive /tmp/db 8080
	
First Parameter is the path to the database (default is `./db`) and second is the port of the webserver (default is `80`).

Golive persitance is based on [Bolt](https://github.com/boltdb/bolt), a pure Go key/value store. No need to install an extra database. Dead simple I was saying !

## Docker

An image is available here : https://hub.docker.com/r/alexandrejomin/golive/

Work in progress.

## API

### `POST` /api/deliveries

**payload :**

```
{
    "software": "delta",
    "version": "0.1.3",
    "environment": "production",
    "repository": "github.com/foo/my-awesome-backend"
}
```

### `GET` /api/deliveries
### `GET` /api/deliveries/:id
### `GET` /api/heatmap


CHANGELOG.md
### fix(dao): GetBetweenInterval nows returns data even if max is not found (HEAD -> master)
>Tue, 20 Dec 2016 16:12:56 +0100

>Author: Alexandre Jomin (alexandrejomin@gmail.com)

>Commiter: Alexandre Jomin (alexandrejomin@gmail.com)




### feat(dao,handler): add limit and offset to GetAll, handler now requests for deliveries in a 30 days time lapse (feat/limitAndOffsetOnDAO)
>Mon, 19 Dec 2016 12:36:43 +0100

>Author: Alexandre Jomin (alexandrejomin@gmail.com)

>Commiter: Alexandre Jomin (alexandrejomin@gmail.com)




### refactor(dao): better iterating over keys in GetAll
>Sat, 17 Dec 2016 20:01:46 +0100

>Author: Alexandre Jomin (alexandrejomin@gmail.com)

>Commiter: Alexandre Jomin (alexandrejomin@gmail.com)




### docs(readme): add features list and a link to docker image
>Wed, 14 Dec 2016 08:08:13 +0100

>Author: Alexandre Jomin (alexandrejomin@gmail.com)

>Commiter: Alexandre Jomin (alexandrejomin@gmail.com)




### feat(frontend): format the date of deployment with moment.js
>Tue, 13 Dec 2016 20:16:06 +0100

>Author: Alexandre Jomin (alexandrejomin@gmail.com)

>Commiter: Alexandre Jomin (alexandrejomin@gmail.com)




### chore: add build status image
>Thu, 8 Dec 2016 14:06:34 +0100

>Author: Alexandre Jomin (alexandrejomin@gmail.com)

>Commiter: Alexandre Jomin (alexandrejomin@gmail.com)




### chore: add travis CI (feat/travis)
>Thu, 8 Dec 2016 12:09:27 +0100

>Author: Alexandre Jomin (alexandrejomin@gmail.com)

>Commiter: Alexandre Jomin (alexandrejomin@gmail.com)




### test(handler) checking api endpoints
>Thu, 8 Dec 2016 12:08:32 +0100

>Author: Alexandre Jomin (alexandrejomin@gmail.com)

>Commiter: Alexandre Jomin (alexandrejomin@gmail.com)




### docs(readme): fix a typo
>Wed, 7 Dec 2016 20:27:42 +0100

>Author: Alexandre Jomin (alexandrejomin@gmail.com)

>Commiter: Alexandre Jomin (alexandrejomin@gmail.com)




### feat(frontend): heatmap dynamically displays last 12 monthes
>Wed, 7 Dec 2016 19:39:56 +0100

>Author: Alexandre Jomin (alexandrejomin@gmail.com)

>Commiter: Alexandre Jomin (alexandrejomin@gmail.com)




### Very first commit ヅ
>Sun, 27 Nov 2016 22:06:25 +0100

>Author: Alexandre Jomin (alexandrejomin@gmail.com)

>Commiter: Alexandre Jomin (alexandrejomin@gmail.com)






data/bolt_test.go
package data

import "testing"

var s Source

func getBoltSource() (Source, error) {

	var err error

	if s != nil {
		return s, nil
	}

	s, err = NewBoltSource("/tmp/db")
	return s, err
}

func TestNewBoltSource(t *testing.T) {
	_, err := getBoltSource()
	if err != nil {
		t.Fatal(err)
	}
}

func TestInsert(t *testing.T) {

	b, err := getBoltSource()

	if err != nil {
		t.Fatal(err)
	}

	err = b.Insert("Foo", []string{"foo", "bar"})

	if err != nil {
		t.Fatal(err)
	}
}


data/utils_test.go
package data

import (
	"testing"
	"time"
)

func TestGetDateFromNbOfDays(t *testing.T) {

	d, err := GetDateFromNbOfDays(-30)

	if err != nil {
		t.Fatal(err)
	}

	e := d.AddDate(0, 0, 30).Truncate(time.Minute)
	n := time.Now().Truncate(time.Minute).UTC()

	// Check if we have the same time
	if !e.Equal(n) {
		t.Fatalf("Date should be equal ! given : `%s` but expecting `%s`", e, n)
	}

}


handler/delivery_test.go
package handler

import (
	"bytes"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/alexjomin/golive/data"
	"github.com/alexjomin/golive/model"
)

var h *Handler
var id string

func createTestHandler() (*Handler, error) {

	if h != nil {
		return h, nil
	}

	testConfig := map[string]string{
		"path": "/tmp/db",
		"port": "8080",
	}

	s, err := data.NewSource(data.SourceBolt, testConfig)

	if err != nil {
		return nil, err
	}

	h = &Handler{
		Source: s,
	}

	return h, nil
}

// TestPostDocument test to create a document
func TestPostDocument(t *testing.T) {

	h, err := createTestHandler()

	if err != nil {
		t.Fatal(err)
	}

	payload := model.Delivery{
		Software:    "my-awesome-api",
		Version:     "0.1.7",
		Environment: "production",
		Repository:  "github.com/foo/my-awesome-backend",
	}

	body, err := json.Marshal(payload)

	req, err := http.NewRequest(http.MethodPost, "/api/deliveries", bytes.NewBuffer(body))

	if err != nil {
		t.Fatal(err)
	}

	rr := httptest.NewRecorder()
	handler := http.HandlerFunc(h.Create)
	handler.ServeHTTP(rr, req)

	if status := rr.Code; status != http.StatusCreated {
		t.Errorf("handler returned wrong status code: got %v want %v", status, http.StatusCreated)
	}

	if rr.Body == nil {
		t.Fatal("No body in the response")
	}

	responseBody := &model.Delivery{}
	err = json.Unmarshal(rr.Body.Bytes(), responseBody)

	if err != nil {
		t.Fatal(err)
	}

	if responseBody.Date == "" {
		t.Fatal("responseBody.date is empty")
	}

	// Set The Id
	id = responseBody.Date
	t.Logf("Id of the document is %s", id)

}

// TestGetDocument retrieve all documents
func TestGetAllDocuments(t *testing.T) {

	h, err := createTestHandler()

	if err != nil {
		t.Fatal(err)
	}

	req, err := http.NewRequest(http.MethodGet, "/api/deliveries", nil)

	if err != nil {
		t.Fatal(err)
	}

	rr := httptest.NewRecorder()
	handler := http.HandlerFunc(h.Get)
	handler.ServeHTTP(rr, req)

	if status := rr.Code; status != http.StatusOK {
		t.Errorf("handler returned wrong status code: got %v want %v", status, http.StatusOK)
	}

}

// TestDeleteADocument deletea a document
func TestDeleteADocument(t *testing.T) {

	h, err := createTestHandler()

	if err != nil {
		t.Fatal(err)
	}

	req, err := http.NewRequest(http.MethodDelete, "/api/deliveries/"+id, nil)

	if err != nil {
		t.Fatal(err)
	}

	rr := httptest.NewRecorder()
	handler := http.HandlerFunc(h.Delete)
	handler.ServeHTTP(rr, req)

	if status := rr.Code; status != http.StatusNoContent {
		t.Errorf("handler returned wrong status code: got %v want %v", status, http.StatusNoContent)
	}

}


main.go
package main

import (
	"log"
	"net/http"
	"os"

	"github.com/alexjomin/golive/data"
	"github.com/alexjomin/golive/handler"
	"github.com/gorilla/mux"
)

func main() {

	log.SetOutput(os.Stdout)

	config := map[string]string{}
	config["path"] = "./db"

	port := "80"

	if len(os.Args) > 1 {
		config["path"] = os.Args[1]
	}

	if len(os.Args) > 2 {
		port = os.Args[2]
	}

	s, err := data.NewSource(data.SourceBolt, config)

	if err != nil {
		log.Fatalf("Can't get a data source at startup, %s", err)
	}

	h := &handler.Handler{
		Source: s,
	}

	r := mux.NewRouter()

	r.HandleFunc("/api/deliveries", h.Get).Methods(http.MethodGet)
	r.HandleFunc("/api/deliveries/{id}", h.Delete).Methods(http.MethodDelete)
	r.HandleFunc("/api/deliveries", h.Create).Methods(http.MethodPost)
	r.HandleFunc("/api/heatmap", h.Heatmap).Methods(http.MethodGet)

	r.PathPrefix("/").Handler(http.StripPrefix("/", http.FileServer(http.Dir("./static"))))

	log.Printf("Path of database : %s", config["path"])
	log.Printf("Start http server on port : %s", port)
	log.Fatal(http.ListenAndServe(":"+port, r))
}


data/bolt.go
package data

import (
	"bytes"
	"encoding/json"
	"fmt"

	"github.com/alexjomin/golive/model"
	"github.com/boltdb/bolt"
)

type boltSource struct {
	db     *bolt.DB
	bucket []byte
}

const bucketName = "deliveries"

// NewBoltSource returns an bolt based data source
func NewBoltSource(path string) (Source, error) {

	db, err := bolt.Open(path, 0600, nil)

	if err != nil {
		return nil, err
	}

	err = db.Update(func(tx *bolt.Tx) error {
		_, err := tx.CreateBucketIfNotExists([]byte(bucketName))

		if err != nil {
			return err
		}

		return nil
	})

	if err != nil {
		return nil, err
	}

	return &boltSource{
		db:     db,
		bucket: []byte(bucketName),
	}, nil

}

func (s *boltSource) Get(key string) (interface{}, error) {

	var v []byte

	_ = s.db.View(func(tx *bolt.Tx) error {
		b := tx.Bucket(s.bucket)

		if b == nil {
			fmt.Println("bucket is nil")
			return nil
		}
		v = b.Get([]byte(key))
		return nil
	})

	if v == nil {
		return nil, nil
	}

	return v, nil

}

func (s *boltSource) Delete(key string) error {
	return s.db.Update(func(tx *bolt.Tx) error {
		b := tx.Bucket([]byte(s.bucket))
		return b.Delete([]byte(key))
	})
}

func (s *boltSource) GetAll(offset, limit int) ([]model.Delivery, error) {

	p := []model.Delivery{}

	err := s.db.View(func(tx *bolt.Tx) error {

		c := tx.Bucket([]byte(s.bucket)).Cursor()

		i := 0

		for k, v := c.Last(); k != nil; k, v = c.Prev() {
			if offset > 0 && i < offset {
				i++
				continue
			}

			if limit > 0 && limit == len(p) {
				break
			}
			var d model.Delivery
			json.Unmarshal(v, &d)
			p = append(p, d)
			i++
		}

		return nil

	})

	return p, err

}

func (s *boltSource) GetBetweenInterval(from, to string) ([]model.Delivery, error) {
	p := []model.Delivery{}

	err := s.db.View(func(tx *bolt.Tx) error {

		c := tx.Bucket([]byte(s.bucket)).Cursor()

		min := []byte(from)
		max := []byte(to)

		// Iterate over the time window
		for k, v := c.Last(); k != nil; k, v = c.Prev() {
			if !(bytes.Compare(k, min) >= 0 && bytes.Compare(k, max) <= 0) {
				continue
			}
			var d model.Delivery
			json.Unmarshal(v, &d)
			if d.Date == "" {
				continue
			}
			p = append(p, d)
		}

		return nil
	})

	return p, err

}

func (s *boltSource) Insert(key string, value interface{}) error {

	buf, err := json.Marshal(value)

	if err != nil {
		return err
	}

	_ = s.db.Update(func(tx *bolt.Tx) error {
		b := tx.Bucket(s.bucket)
		err := b.Put([]byte(key), buf)
		return err
	})

	return nil
}


static/app.js
// CalHeatmap
var cal = new CalHeatMap();
var clickDate = null;
var windowWidth = 0;
var resizeTimeout = 0;

function monthAgo(month) {
  date = new Date();
  date.setMonth(date.getMonth() - (month-1), 1);
  return date;
}

function calInit() {
  
  var width = window.innerWidth || document.documentElement.clientWidth || document.getElementsByTagName('body')[0].clientWidth;
  
  if (width != windowWidth) {
    windowWidth = width;
  } else {
    return false;
  }
  
  var months = 12;
  
  if (width < 450) {
    months = 5;
  } else if (width < 520) {
    months = 6;
  } else if (width < 600) {
    months = 7;
  } else if (width < 768) {
    months = 8;
  } else if (width < 992) {
    months = 10;
  }
  
  if (cal._completed) {
    cal.destroy();
    cal = new CalHeatMap();
  }
  
  setTimeout(function() {
    cal.init({
      animationDuration: 0,
      data: "/api/heatmap",
      domain: "month",
      start: monthAgo(months),
      domainGutter: 10,
      range: months,
      legend: [1, 3, 5, 10, 20],
      onClick: function(date, nb) {
        
        if (clickDate !== null && clickDate.toDateString() == date.toDateString()) {
          clickDate = null;
        } else {
          clickDate = date;
        }
        
        vue.filterDate(clickDate);
        
      }
    });
  }, 1);
  
}
calInit();

window.onresize = function() {
  
  if (resizeTimeout) {
    clearTimeout(resizeTimeout);
  }
  resizeTimeout = setTimeout(function() {
      calInit();
  }, 500);
  
};

// Set moment locale
moment.locale(navigator.language || navigator.userLanguage);

// Vue
var apiURL = '/api';

var vue = new Vue({

  el: '#timeline',

  data: {
    data: [],
    deliveries: [],
    softwares: [],
    environments: [],
    currentSoft: "",
    currentEnv: "",
    colors: [
      '#2ecc71',
      '#5bc0de',
      '#3498db'
    ]
  },

  created: function () {
    this.fetchData();
  },

  watch: {
    //currentBranch: 'fetchData'
  },

  filters: {
    truncate: function (v) {
      var newline = v.indexOf('\n');
      return newline > 0 ? v.slice(0, newline) : v;
    },
    formatDate: function (v) {
      return moment(v).format('LLLL');
    },
    getChangelogURL: function(v) {
      if(v.repository.length === 0) return;
      return `https://${v.repository}/src/HEAD/CHANGELOG.md?at=master`;
    }
  },

  methods: {
    populateDeliveries: function(data) {
      
      this.deliveries = [];
      
      if (data.length) {
        for (var i in data) {
          if (
              (this.currentSoft === "" || data[i].software == this.currentSoft) &&
              (this.currentEnv === "" || data[i].environment == this.currentEnv)
            ) {
            this.deliveries.push(data[i]);
          }
        }
      }
      
    },
    filter: function(name){
      
      this.deliveries = [];
      
      if (this.currentSoft == name){
        this.currentSoft = "";
      } else {
        this.currentSoft = name;
      }
      
      this.populateDeliveries(this.data);
      
    },
    filterEnv: function(name){
      
      this.deliveries = [];
      
      if (this.currentEnv == name){
        this.currentEnv = "";
      } else {
        this.currentEnv = name;
      }
      
      this.populateDeliveries(this.data);
      
    },
    filterDate: function(date){
      
      if (date === null) {
        this.populateDeliveries(this.data);
        return;
      }
      
      date = date.toDateString();
      var data = [];
      for (var i in this.data) {
        if ((new Date(this.data[i].date)).toDateString() == date) {
          data.push(this.data[i]);
        }
      }
      
      this.populateDeliveries(data);
      
    },
    fetchData: function () {
      var xhr = new XMLHttpRequest();
      var self = this;
      xhr.open('GET', apiURL + "/deliveries");
      xhr.onload = function () {
        d = JSON.parse(xhr.responseText);
        self.data = d;
        var index, i;
        
        for (i = 0; i < d.length ; i++) {
          
          index = self.softwares.indexOf(d[i].software);
          if (index === -1) {
            self.softwares.push(d[i].software);
            index = self.softwares.length-1;
          }
          
          index = self.environments.indexOf(d[i].environment);
          if (index === -1) {
            self.environments.push(d[i].environment);
            index = self.environments.length-1;
          }
          
        }
        
        self.softwares.sort();
        self.environments.sort();
        
        for (i in self.data) {
          
          index = self.environments.indexOf(self.data[i].environment);
          self.data[i].environmentColor = self.colors[index];
          
        }
        
        self.populateDeliveries(self.data);
      };
      xhr.send();
    }
  }
});

handler/delivery.go
package handler

import (
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/alexjomin/golive/data"
	"github.com/alexjomin/golive/model"
	"github.com/gorilla/mux"
)

// Handler is a custom handler
type Handler struct {
	Source data.Source
}

// Error The HTTP error payload
type Error struct {
	Message string
}

// Delete remove an entry with given key
func (h *Handler) Delete(w http.ResponseWriter, r *http.Request) {

	p := mux.Vars(r)
	id := p["id"]

	err := h.Source.Delete(id)

	if err != nil {
		e := Error{
			Message: fmt.Sprintf("An error occured when trying to delete the delivery : %s", err),
		}
		JSONWithHTTPCode(w, e, 500)
		return
	}

	JSONWithHTTPCode(w, nil, 204)

}

// Create a new delivery
func (h *Handler) Create(w http.ResponseWriter, r *http.Request) {

	decoder := json.NewDecoder(r.Body)
	defer r.Body.Close()

	var t model.Delivery
	err := decoder.Decode(&t)

	if err != nil {
		e := Error{
			Message: "An error has occured during the parsing of the body",
		}
		JSONWithHTTPCode(w, e, 400)
		return
	}

	if t.Software == "" {
		e := Error{
			Message: "The software is mandatory",
		}
		JSONWithHTTPCode(w, e, 400)
		return
	}

	if t.Version == "" {
		e := Error{
			Message: "The version is mandatory",
		}
		JSONWithHTTPCode(w, e, 400)
		return
	}

	n := time.Now().UTC().Format(time.RFC3339)
	t.Date = n

	err = h.Source.Insert(n, t)

	if err != nil {
		e := Error{
			Message: "An error occured when trying to create the delivery",
		}
		JSONWithHTTPCode(w, e, 500)
		return
	}

	JSONWithHTTPCode(w, t, 201)

}

// Get returns all the deliveries
func (h *Handler) Get(w http.ResponseWriter, r *http.Request) {

	from, err := data.GetDateFromNbOfDays(-30)

	if err != nil {
		e := Error{
			Message: "An error occured when trying to retrieve the deliveries",
		}
		JSONWithHTTPCode(w, e, 500)
		return
	}

	*from = from.UTC()
	n := time.Now().UTC().Truncate(time.Minute).Format(time.RFC3339)

	data, err := h.Source.GetBetweenInterval(from.Format(time.RFC3339), n)

	if err != nil {
		e := Error{
			Message: "An error occured when trying to retrieve the deliveries",
		}
		JSONWithHTTPCode(w, e, 500)
		return
	}

	JSON(w, data)

}

// Heatmap returns a list of timestamp where delivery occurs
func (h *Handler) Heatmap(w http.ResponseWriter, r *http.Request) {

	data, err := h.Source.GetAll(0, 0)

	if err != nil {
		e := Error{
			Message: "An error occured when trying to retrieve the deliveries",
		}
		JSONWithHTTPCode(w, e, 500)
		return
	}

	layout := "2006-01-02T15:04:05Z"

	rtr := map[string]int{}

	for _, d := range data {
		t, _ := time.Parse(layout, d.Date)
		ts := fmt.Sprintf("%d", t.Unix())
		rtr[ts] = 1
	}

	JSON(w, rtr)

}

// JSON Outputs a JSON
func JSON(w http.ResponseWriter, d interface{}) {
	JSONWithHTTPCode(w, d, http.StatusOK)
}

// JSONWithHTTPCode Json Output with an HTTP code
func JSONWithHTTPCode(w http.ResponseWriter, d interface{}, code int) {
	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
	w.WriteHeader(code)
	if d != nil {
		err := json.NewEncoder(w).Encode(d)
		if err != nil {
			// panic will cause the http.StatusInternalServerError to be send to users
			panic(err)
		}
	}
}
